File: server/internal/api/problem/problem.go
// Package problem defines Lumilio's language-neutral RFC 9457 vocabulary.
//
// It is intentionally a leaf package: handlers and transports may depend on
// these descriptors, while domain packages remain unaware of HTTP problem URIs.
package problem

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"sort"
)

const (
	MediaType = "application/problem+json"
	Namespace = "https://lumilio.org/problems/"
)

// Descriptor is the canonical definition of one public Problem type.
type Descriptor struct {
	Type       string
	Status     int
	Title      string
	Definition string
	Recovery   string
	Extensions []Extension
}

// Extension documents one bounded, type-specific public member.
type Extension struct {
	Name        string
	Schema      string
	Description string
}

var (
	InvalidCredentials = descriptor("auth/invalid-credentials", http.StatusUnauthorized,
		"Invalid credentials", "The supplied sign-in credentials were not accepted.")
	AuthenticationRequired = descriptor("auth/authentication-required", http.StatusUnauthorized,
		"Authentication required", "A valid authenticated session is required for this operation.")
	SessionExpired = descriptor("auth/session-expired", http.StatusUnauthorized,
		"Session expired", "The browser session can no longer be refreshed.")
	PasskeyUnavailable = descriptor("auth/passkey-unavailable", http.StatusBadRequest,
		"Passkey unavailable", "Passkey authentication is unavailable for this request.")
	MFAInvalid = descriptor("auth/mfa-invalid", http.StatusUnauthorized,
		"Invalid multi-factor authentication", "The supplied multi-factor authentication proof was not accepted.")
	PermissionDenied = descriptor("auth/permission-denied", http.StatusForbidden,
		"Permission denied", "The authenticated user cannot perform this operation.")
	UntrustedOrigin = descriptor("auth/untrusted-origin", http.StatusForbidden,
		"Untrusted request origin", "The browser request origin is not trusted for a session-changing operation.")
	RateLimited = Descriptor{
		Type: Namespace + "auth/rate-limited", Status: http.StatusTooManyRequests,
		Title: "Too many requests", Definition: "Authentication requests are temporarily rate limited.",
		Recovery:   "Wait for Retry-After, then retry the authentication request.",
		Extensions: []Extension{{Name: "retry_after_seconds", Schema: "integer", Description: "Whole seconds until the request may be retried."}},
	}
	AppNotInitialized = descriptor("bootstrap/app-not-initialized", http.StatusConflict,
		"Application not initialized", "First-run setup must finish before this operation is available.")
	RepositoryUnavailable = descriptor("repository/unavailable", http.StatusServiceUnavailable,
		"Repository unavailable", "The Repository required by the operation is unavailable.")
	RepositoryConflict = Descriptor{
		Type: Namespace + "repository/conflict", Status: http.StatusConflict,
		Title: "Repository conflict", Definition: "Repository state requires an explicit recovery choice.",
		Recovery: "Choose one of the response's allowed recovery actions, then retry.",
		Extensions: []Extension{
			{Name: "conflict_type", Schema: "string", Description: "Stable recovery classification."},
			{Name: "repository_id", Schema: "string", Description: "Safe Repository identity, when known."},
			{Name: "actions", Schema: "array[string]", Description: "Stable recovery actions allowed for this conflict."},
		},
	}
	StorageConfirmationRequired = descriptor("storage/confirmation-required", http.StatusConflict,
		"Storage confirmation required", "The storage operation requires explicit administrator risk confirmation.")
	ServiceUnavailable = descriptor("service/unavailable", http.StatusServiceUnavailable,
		"Service unavailable", "A required optional service is currently unavailable.")
	InvalidMediaRequest = descriptor("media/invalid-request", http.StatusUnprocessableEntity,
		"Invalid media request", "The requested media operation cannot process the supplied media.")
	ImageEmbeddingMissing = descriptor("media/image-embedding-missing", http.StatusConflict,
		"Image embedding missing", "The selected asset does not have an Image Semantic Analysis embedding.")
	SemanticAnalysisUnavailable = descriptor("lumen/image-semantic-analysis-unavailable", http.StatusServiceUnavailable,
		"Image Semantic Analysis unavailable", "Image Semantic Analysis is not currently available.")
	AgentFailed = descriptor("agent/operation-failed", http.StatusInternalServerError,
		"Agent operation failed", "The Agent operation could not be completed.")
	UploadProcessingFailed = descriptor("upload/processing-failed", http.StatusInternalServerError,
		"Upload processing failed", "An accepted upload could not be processed.")
	RepositoryScanFailed = descriptor("repository/scan-failed", http.StatusInternalServerError,
		"Repository scan failed", "An accepted Repository scan could not be completed.")
	RepositoryScanIncomplete = descriptor("repository/scan-incomplete", http.StatusInternalServerError,
		"Repository scan incomplete", "A Repository scan completed with partial results.")
	BackupRestoreFailed = descriptor("backup/restore-failed", http.StatusInternalServerError,
		"Backup restore failed", "An accepted database restore could not be completed.")
	HostActionFailed = descriptor("storage/host-action-failed", http.StatusInternalServerError,
		"Native host action failed", "An accepted native host action could not be completed.")
	HostActionExpired = descriptor("storage/host-action-expired", http.StatusInternalServerError,
		"Native host action expired", "A native host action expired before it was approved.")
	CloudImportFailed = descriptor("cloud/import-failed", http.StatusInternalServerError,
		"Cloud import failed", "An accepted cloud import could not be completed.")
)

var registry = mustRegistry(
	InvalidCredentials,
	AuthenticationRequired,
	SessionExpired,
	PasskeyUnavailable,
	MFAInvalid,
	PermissionDenied,
	UntrustedOrigin,
	RateLimited,
	AppNotInitialized,
	RepositoryUnavailable,
	RepositoryConflict,
	StorageConfirmationRequired,
	ServiceUnavailable,
	InvalidMediaRequest,
	ImageEmbeddingMissing,
	SemanticAnalysisUnavailable,
	AgentFailed,
	UploadProcessingFailed,
	RepositoryScanFailed,
	RepositoryScanIncomplete,
	BackupRestoreFailed,
	HostActionFailed,
	HostActionExpired,
	CloudImportFailed,
)

func descriptor(path string, status int, title, definition string) Descriptor {
	return Descriptor{
		Type: Namespace + path, Status: status, Title: title, Definition: definition,
		Recovery: "Correct the documented condition, then retry the operation.",
	}
}

func mustRegistry(descriptors ...Descriptor) map[string]Descriptor {
	result := make(map[string]Descriptor, len(descriptors))
	for _, item := range descriptors {
		if item.Type == "" || item.Status < 400 || item.Status > 599 || item.Title == "" || item.Definition == "" || item.Recovery == "" {
			panic(fmt.Sprintf("invalid Problem descriptor: %#v", item))
		}
		if _, exists := result[item.Type]; exists {
			panic("duplicate Problem descriptor: " + item.Type)
		}
		result[item.Type] = item
	}
	return result
}

// Lookup returns the registered descriptor for a Lumilio Problem type.
func Lookup(problemType string) (Descriptor, bool) {
	descriptor, ok := registry[problemType]
	return descriptor, ok
}

// Registered returns the canonical catalog in stable URI order.
func Registered() []Descriptor {
	result := make([]Descriptor, 0, len(registry))
	for _, descriptor := range registry {
		result = append(result, descriptor)
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Type < result[j].Type })
	return result
}

// Details is Lumilio's deliberately narrow RFC 9457 base response.
type Details struct {
	Type     string `json:"type" binding:"required" example:"https://lumilio.org/problems/auth/invalid-credentials"`
	Status   int    `json:"status" binding:"required" example:"401"`
	Instance string `json:"instance" binding:"required" example:"urn:lumilio:problem:0123456789abcdef0123456789abcdef"`
}

// RateLimitedDetails mirrors Retry-After for localized countdown copy.
type RateLimitedDetails struct {
	Details
	RetryAfterSeconds int64 `json:"retry_after_seconds" binding:"required" minimum:"1" example:"60"`
}

// RepositoryConflictDetails carries only bounded recovery facts. Host paths
// and display prose are intentionally absent.
type RepositoryConflictDetails struct {
	Details
	ConflictType string   `json:"conflict_type" binding:"required" example:"repository_identity"`
	RepositoryID string   `json:"repository_id,omitempty" example:"550e8400-e29b-41d4-a716-446655440000"`
	Actions      []string `json:"actions,omitempty" example:"relocate,copy"`
}

// Reference is the transport-neutral equivalent used by streams and
// persisted asynchronous operation DTOs. It intentionally has no HTTP status.
type Reference struct {
	Type      string `json:"type" binding:"required" example:"https://lumilio.org/problems/upload/processing-failed"`
	Instance  string `json:"instance" binding:"required" example:"urn:lumilio:problem:0123456789abcdef0123456789abcdef"`
	Retryable *bool  `json:"retryable,omitempty"`
}

// NewInstance creates an opaque occurrence URI for one request or ephemeral
// stream failure. It contains no caller or domain identifier.
func NewInstance() string {
	var value [16]byte
	if _, err := rand.Read(value[:]); err != nil {
		panic("generate Problem occurrence ID: " + err.Error())
	}
	return instanceFromBytes(value[:])
}

// StableInstance derives an opaque occurrence URI from a durable operation
// identity. The identity is not embedded in the URI; the same persisted
// operation therefore retains its support reference across restarts.
func StableInstance(descriptor Descriptor, operationIdentity string) string {
	digest := sha256.Sum256([]byte("lumilio.problem.instance.v1\x00" + descriptor.Type + "\x00" + operationIdentity))
	return instanceFromBytes(digest[:16])
}

// NewReference creates a Problem Reference for an ephemeral stream or item
// failure. References are limited to descriptors in the closed registry.
func NewReference(descriptor Descriptor, retryable bool) Reference {
	return reference(descriptor, NewInstance(), retryable)
}

// ReferenceFor creates a stable Problem Reference for a persisted operation.
func ReferenceFor(descriptor Descriptor, operationIdentity string, retryable bool) Reference {
	return reference(descriptor, StableInstance(descriptor, operationIdentity), retryable)
}

func reference(descriptor Descriptor, instance string, retryable bool) Reference {
	registered, ok := Lookup(descriptor.Type)
	if !ok || registered.Status != descriptor.Status {
		registered = ServiceUnavailable
	}
	return Reference{Type: registered.Type, Instance: instance, Retryable: &retryable}
}

func instanceFromBytes(value []byte) string {
	return "urn:lumilio:problem:" + hex.EncodeToString(value)
}

type bodyKind uint8

const (
	baseBody bodyKind = iota
	rateLimitedBody
	repositoryConflictBody
)

// Failure pairs one safe public descriptor with its private diagnostic cause.
// Its fields are private so callers cannot add arbitrary response members.
type Failure struct {
	descriptor        Descriptor
	status            int
	cause             error
	kind              bodyKind
	retryAfterSeconds int64
	conflictType      string
	repositoryID      string
	actions           []string
}

// About creates a generic about:blank failure for a status whose meaning and
// operation-local UI fallback are sufficient.
func About(status int, cause error) Failure {
	if status < 400 || status > 599 {
		status = http.StatusInternalServerError
	}
	return Failure{descriptor: Descriptor{Type: "about:blank", Status: status}, status: status, cause: cause}
}

// New creates a registered Problem failure without extension members.
func New(descriptor Descriptor, cause error) Failure {
	registered, ok := Lookup(descriptor.Type)
	if !ok || registered.Status != descriptor.Status || len(registered.Extensions) != 0 {
		return About(http.StatusInternalServerError, fmt.Errorf("invalid Problem descriptor %q: %w", descriptor.Type, cause))
	}
	return Failure{descriptor: registered, status: registered.Status, cause: cause}
}

// NewRateLimited creates the exact typed rate-limit subtype.
func NewRateLimited(cause error, retryAfterSeconds int64) Failure {
	if retryAfterSeconds < 1 {
		retryAfterSeconds = 1
	}
	return Failure{descriptor: RateLimited, status: RateLimited.Status, cause: cause, kind: rateLimitedBody, retryAfterSeconds: retryAfterSeconds}
}

// NewRepositoryConflict creates the exact typed Repository conflict subtype.
func NewRepositoryConflict(cause error, conflictType, repositoryID string, actions []string) Failure {
	return Failure{
		descriptor: RepositoryConflict, status: RepositoryConflict.Status, cause: cause,
		kind: repositoryConflictBody, conflictType: conflictType, repositoryID: repositoryID,
		actions: append([]string(nil), actions...),
	}
}

func (f Failure) Status() int            { return f.status }
func (f Failure) Type() string           { return f.descriptor.Type }
func (f Failure) Cause() error           { return f.cause }
func (f Failure) Descriptor() Descriptor { return f.descriptor }

// Body materializes the public typed payload after the transport assigns an
// opaque occurrence instance.
func (f Failure) Body(instance string) any {
	base := Details{Type: f.Type(), Status: f.Status(), Instance: instance}
	switch f.kind {
	case rateLimitedBody:
		return RateLimitedDetails{Details: base, RetryAfterSeconds: f.retryAfterSeconds}
	case repositoryConflictBody:
		return RepositoryConflictDetails{
			Details: base, ConflictType: f.conflictType, RepositoryID: f.repositoryID,
			Actions: append([]string(nil), f.actions...),
		}
	default:
		return base
	}
}

File: web/src/lib/http-commons/problem.ts
import type { TFunction } from "i18next";
import type { components } from "./schema";

export type KnownProblem = components["schemas"]["api.ProblemResponse"];
export type KnownProblemType = KnownProblem["type"];
export type KnownProblemReference = components["schemas"]["api.ProblemReference"];

const INSTANCE_PATTERN = /^urn:lumilio:problem:[0-9a-f]{32}$/;

const KNOWN_TYPES = {
  "about:blank": true,
  "https://lumilio.org/problems/agent/operation-failed": true,
  "https://lumilio.org/problems/auth/authentication-required": true,
  "https://lumilio.org/problems/auth/invalid-credentials": true,
  "https://lumilio.org/problems/auth/mfa-invalid": true,
  "https://lumilio.org/problems/auth/passkey-unavailable": true,
  "https://lumilio.org/problems/auth/permission-denied": true,
  "https://lumilio.org/problems/auth/rate-limited": true,
  "https://lumilio.org/problems/auth/session-expired": true,
  "https://lumilio.org/problems/auth/untrusted-origin": true,
  "https://lumilio.org/problems/backup/restore-failed": true,
  "https://lumilio.org/problems/bootstrap/app-not-initialized": true,
  "https://lumilio.org/problems/cloud/import-failed": true,
  "https://lumilio.org/problems/lumen/image-semantic-analysis-unavailable": true,
  "https://lumilio.org/problems/media/image-embedding-missing": true,
  "https://lumilio.org/problems/media/invalid-request": true,
  "https://lumilio.org/problems/repository/conflict": true,
  "https://lumilio.org/problems/repository/scan-failed": true,
  "https://lumilio.org/problems/repository/scan-incomplete": true,
  "https://lumilio.org/problems/repository/unavailable": true,
  "https://lumilio.org/problems/service/unavailable": true,
  "https://lumilio.org/problems/storage/confirmation-required": true,
  "https://lumilio.org/problems/storage/host-action-expired": true,
  "https://lumilio.org/problems/storage/host-action-failed": true,
  "https://lumilio.org/problems/upload/processing-failed": true,
} as const satisfies Record<KnownProblemType, true>;

export type NormalizedProblem =
  | {
      kind: "problem";
      type: string;
      status: number;
      instance: string;
      retryAfterSeconds?: number;
      conflictType?: string;
      repositoryID?: string;
      actions?: readonly string[];
    }
  | { kind: "abort" }
  | { kind: "network" }
  | { kind: "malformed" };

export type NormalizedProblemReference = {
  type: string;
  instance: string;
  retryable?: boolean;
  retryAfterSeconds?: number;
  conflictType?: string;
  repositoryID?: string;
  actions?: readonly string[];
};

function record(value: unknown): Record<string, unknown> | undefined {
  if (!value || typeof value !== "object" || Array.isArray(value)) return undefined;
  return value as Record<string, unknown>;
}

function safeExtensions(source: Record<string, unknown>) {
  const retryAfterSeconds =
    Number.isInteger(source.retry_after_seconds) && Number(source.retry_after_seconds) > 0
      ? Number(source.retry_after_seconds)
      : undefined;
  const conflictType =
    typeof source.conflict_type === "string" && source.conflict_type.length <= 128
      ? source.conflict_type
      : undefined;
  const repositoryID =
    typeof source.repository_id === "string" && source.repository_id.length <= 128
      ? source.repository_id
      : undefined;
  const actions =
    Array.isArray(source.actions) &&
    source.actions.length <= 16 &&
    source.actions.every((action) => typeof action === "string" && action.length <= 64)
      ? (source.actions as string[]).slice()
      : undefined;
  return { retryAfterSeconds, conflictType, repositoryID, actions };
}

export function normalizeProblem(value: unknown): NormalizedProblem {
  if (isNormalizedProblem(value)) return value;
  if (value instanceof DOMException && value.name === "AbortError") return { kind: "abort" };
  if (value instanceof TypeError) return { kind: "network" };

  const source = record(value);
  if (!source) return { kind: "malformed" };
  if (
    typeof source.type !== "string" ||
    !Number.isInteger(source.status) ||
    Number(source.status) < 400 ||
    Number(source.status) > 599 ||
    typeof source.instance !== "string" ||
    !INSTANCE_PATTERN.test(source.instance)
  ) {
    return { kind: "malformed" };
  }

  return {
    kind: "problem",
    type: source.type,
    status: Number(source.status),
    instance: source.instance,
    ...safeExtensions(source),
  };
}

export function normalizeProblemReference(value: unknown): NormalizedProblemReference | undefined {
  const source = record(value);
  if (
    !source ||
    typeof source.type !== "string" ||
    typeof source.instance !== "string" ||
    !INSTANCE_PATTERN.test(source.instance)
  ) {
    return undefined;
  }
  const retryable = typeof source.retryable === "boolean" ? source.retryable : undefined;
  return { type: source.type, instance: source.instance, retryable, ...safeExtensions(source) };
}

export function getProblemType(value: unknown): string | undefined {
  const normalized = normalizeProblem(value);
  return normalized.kind === "problem" ? normalized.type : undefined;
}

export function getProblemInstance(value: unknown): string | undefined {
  const normalized = normalizeProblem(value);
  return normalized.kind === "problem" ? normalized.instance : undefined;
}

export function isAbortProblem(value: unknown): boolean {
  return normalizeProblem(value).kind === "abort";
}

export async function readProblemResponse(response: Response): Promise<NormalizedProblem> {
  if (response.ok) return { kind: "malformed" };
  const contentType = response.headers.get("Content-Type")?.split(";", 1)[0]?.trim();
  if (contentType !== "application/problem+json") return { kind: "malformed" };
  const payload: unknown = await response.json().catch(() => undefined);
  return normalizeProblem(payload);
}

export function localizeProblem(value: unknown, t: TFunction, operationFallback: string): string {
  const normalized = normalizeProblem(value);
  if (normalized.kind !== "problem" || !isKnownType(normalized.type)) return operationFallback;

  return localizeKnownProblem(normalized.type, normalized, t, operationFallback);
}

export function localizeProblemReference(
  value: unknown,
  t: TFunction,
  operationFallback: string,
): string {
  const normalized = normalizeProblemReference(value);
  if (!normalized || !isKnownType(normalized.type)) return operationFallback;
  return localizeKnownProblem(normalized.type, normalized, t, operationFallback);
}

export function localizeAPIProblem(
  value: unknown,
  t: TFunction,
  operationFallback: string,
): string {
  const normalized = normalizeProblem(value);
  if (normalized.kind === "problem") return localizeProblem(normalized, t, operationFallback);
  return localizeProblemReference(value, t, operationFallback);
}

function localizeKnownProblem(
  problemType: KnownProblemType,
  facts: {
    retryAfterSeconds?: number;
    conflictType?: string;
  },
  t: TFunction,
  operationFallback: string,
): string {
  switch (problemType) {
    case "about:blank":
      return operationFallback;
    case "https://lumilio.org/problems/auth/invalid-credentials":
      return t("apiErrors.auth.invalidCredentials", "The username or password is incorrect.");
    case "https://lumilio.org/problems/auth/authentication-required":
      return t("apiErrors.auth.authenticationRequired", "Sign in to continue.");
    case "https://lumilio.org/problems/auth/session-expired":
      return t("apiErrors.auth.sessionExpired", "Your session expired. Sign in again.");
    case "https://lumilio.org/problems/auth/passkey-unavailable":
      return t(
        "apiErrors.auth.passkeyUnavailable",
        "Passkeys are unavailable in this browser or deployment.",
      );
    case "https://lumilio.org/problems/auth/mfa-invalid":
      return t("apiErrors.auth.mfaInvalid", "The verification code is invalid or expired.");
    case "https://lumilio.org/problems/auth/permission-denied":
      return t("apiErrors.auth.permissionDenied", "You do not have permission to do that.");
    case "https://lumilio.org/problems/auth/untrusted-origin":
      return t("apiErrors.auth.untrustedOrigin", "This browser origin is not trusted.");
    case "https://lumilio.org/problems/auth/rate-limited":
      return facts.retryAfterSeconds
        ? t(
            "apiErrors.auth.rateLimitedWithDelay",
            "Too many attempts. Try again in {{count}} seconds.",
            { count: facts.retryAfterSeconds },
          )
        : t("apiErrors.auth.rateLimited", "Too many attempts. Try again later.");
    case "https://lumilio.org/problems/bootstrap/app-not-initialized":
      return t("apiErrors.bootstrap.appNotInitialized", "Finish initial setup before continuing.");
    case "https://lumilio.org/problems/backup/restore-failed":
      return t(
        "apiErrors.backup.restoreFailed",
        "The database restore could not be completed. The previous database remains active.",
      );
    case "https://lumilio.org/problems/cloud/import-failed":
      return t("apiErrors.cloud.importFailed", "The cloud import could not be completed.");
    case "https://lumilio.org/problems/repository/unavailable":
      return t(
        "apiErrors.repository.unavailable",
        "The Repository required for this operation is unavailable.",
      );
    case "https://lumilio.org/problems/repository/conflict":
      return localizeRepositoryConflict(facts.conflictType, t);
    case "https://lumilio.org/problems/repository/scan-failed":
      return t("apiErrors.repository.scanFailed", "The Repository scan could not be completed.");
    case "https://lumilio.org/problems/repository/scan-incomplete":
      return t(
        "apiErrors.repository.scanIncomplete",
        "The Repository scan completed with partial results.",
      );
    case "https://lumilio.org/problems/storage/confirmation-required":
      return t(
        "apiErrors.storage.confirmationRequired",
        "Review and confirm the Storage Location risks before continuing.",
      );
    case "https://lumilio.org/problems/storage/host-action-expired":
      return t(
        "apiErrors.storage.hostActionExpired",
        "The Desktop approval expired. Start the action again.",
      );
    case "https://lumilio.org/problems/storage/host-action-failed":
      return t(
        "apiErrors.storage.hostActionFailed",
        "The Desktop could not complete the storage action.",
      );
    case "https://lumilio.org/problems/service/unavailable":
      return t("apiErrors.service.unavailable", "A required service is unavailable.");
    case "https://lumilio.org/problems/media/invalid-request":
      return t("apiErrors.media.invalidRequest", "The media request could not be processed.");
    case "https://lumilio.org/problems/media/image-embedding-missing":
      return t(
        "apiErrors.media.imageEmbeddingMissing",
        "The selected asset has not completed Image Semantic Analysis.",
      );
    case "https://lumilio.org/problems/lumen/image-semantic-analysis-unavailable":
      return t(
        "apiErrors.lumen.imageSemanticAnalysisUnavailable",
        "Image Semantic Analysis is unavailable.",
      );
    case "https://lumilio.org/problems/agent/operation-failed":
      return t("apiErrors.agent.operationFailed", "The Agent operation could not be completed.");
    case "https://lumilio.org/problems/upload/processing-failed":
      return t("apiErrors.upload.processingFailed", "The upload could not be processed.");
    default:
      return assertNever(problemType);
  }
}

function localizeRepositoryConflict(conflictType: string | undefined, t: TFunction): string {
  switch (conflictType) {
    case "repository_identity":
      return t(
        "apiErrors.repository.conflicts.identity",
        "This Repository identity is already registered.",
      );
    case "existing_repository_found":
      return t(
        "apiErrors.repository.conflicts.existingFound",
        "An existing Repository was found in this Storage Location.",
      );
    case "repository_marker_invalid":
      return t(
        "apiErrors.repository.conflicts.invalidMarker",
        "The selected Repository marker needs attention.",
      );
    case "storage_location_offline":
      return t(
        "apiErrors.repository.conflicts.storageLocationOffline",
        "The Storage Location is offline.",
      );
    default:
      return t(
        "apiErrors.repository.conflicts.generic",
        "The Repository needs attention before continuing.",
      );
  }
}

function isKnownType(value: string): value is KnownProblemType {
  return Object.hasOwn(KNOWN_TYPES, value);
}

function isNormalizedProblem(value: unknown): value is NormalizedProblem {
  const source = record(value);
  return Boolean(
    source &&
    (source.kind === "problem" ||
      source.kind === "abort" ||
      source.kind === "network" ||
      source.kind === "malformed"),
  );
}

function assertNever(value: never): never {
  throw new Error(`Unhandled generated Problem type: ${String(value)}`);
}
